nucliadb/nucliadb/common/maindb/tikv.py

```diff
@@ -218,21 +213,135 @@ async def count(self, match: str) -> int:
         return value
 
 
-class TiKVDriver(Driver):
-    tikv = None
+class TiKVTransaction(Transaction):
+    driver: TiKVDriver
+
+    def __init__(self, txn: Any, driver: TiKVDriver):
+        self.txn = txn
+        self.driver = driver
+        self.data_layer = TiKVDataLayer(txn)
+        self.open = True
+
+    async def abort(self):
+        if not self.open:
+            return
+        await self.data_layer.abort()
+        self.open = False
+
+    async def commit(self):
+        assert self.open
+        await self.data_layer.commit()
+        self.open = False
+
+    async def batch_get(self, keys: list[str]) -> list[Optional[bytes]]:
+        return await self.data_layer.batch_get(keys)
+
+    @backoff.on_exception(backoff.expo, (TimeoutError,), max_tries=2)
+    async def get(self, key: str) -> Optional[bytes]:
+        return await self.data_layer.get(key)
+
+    async def set(self, key: str, value: bytes) -> None:
+        return await self.data_layer.set(key, value)
+
+    async def delete(self, key: str) -> None:
+        return await self.data_layer.delete(key)
+
+    async def keys(
+        self, match: str, count: int = DEFAULT_SCAN_LIMIT, include_start: bool = True
+    ):
+        # XXX must have connection outside of current txn
+        conn_holder = self.driver.get_connection_holder()
+        txn = await conn_holder.get_snapshot()
+        dl = TiKVDataLayer(txn)
+
+        async for key in dl.keys(match, count, include_start):
+            yield key
+
+    async def count(self, match: str) -> int:
+        return await self.data_layer.count(match)
+
+
+class ReadOnlyTiKVTransaction(Transaction):
+    driver: TiKVDriver
+
+    def __init__(self, connection: asynchronous.Snapshot, driver: TiKVDriver):
+        self.connection = connection
+        self.data_layer = TiKVDataLayer(connection)
+        self.driver = driver
+
+    async def abort(self):
```

Comment:
no need to implement abort?

Reply:
Nope, snapshot does not have an abort method.